### ISSUE TITLE
Nested proptypes in FlowType causes RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by willlma** [yannickcr/eslint-plugin-react#913]
I'm guessing this is causing an infinite loop. Not sure is this belongs here or in the [Flow repo](https://github.com/facebook/flow/) (or if I'm just doing things wrong). I've got a nested array that looks like this:

```
type Note = {text: string, children?: Note[]}
type Props = {
  notes: Note[]
};
```

Which is throwing `RangeError: Maximum call stack size exceeded`

Here's the full stack (absolute path to project abbreviated as `~`):

```
    at getKeyValue (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:289:25)
    at iterateProperties (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:308:21)
    at buildTypeAnnotationDeclarationTypes (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:441:11)
    at buildTypeAnnotationDeclarationTypes (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:433:20)
    at buildTypeAnnotationDeclarationTypes (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:472:28)
    at ~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:442:54
    at iterateProperties (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:311:11)
    at buildTypeAnnotationDeclarationTypes (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:441:11)
    at buildTypeAnnotationDeclarationTypes (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:433:20)
    at buildTypeAnnotationDeclarationTypes (~/node_modules/eslint-plugin-react/lib/rules/prop-types.js:472:28)
```

I've stuck with a simple `type Props = { notes: Object[] }` for now

**Comment by ljharb** [comment]
See also #1032.

**Comment by jetpacmonkey** [comment]
Is there any good workaround here? I'm still seeing this issue.

**Comment by ljharb** [comment]
#1138 / #1377 / #1412 should help, but this might still require further fixes.

**Comment by jetpacmonkey** [comment]
It looks like #1138 did address this issue, but it was closed in favor of #1377, which, as far as I can tell, does not.

**Comment by jetpacmonkey** [comment]
I opened #1653 based off of the commit in #1138 that fixed recursive proptype definitions.
